src/main/java/frc/team2412/robot/subsystems/VisionSubsystem.java

```diff
@@ -0,0 +1,128 @@
+package frc.team2412.robot.subsystems;
+
+import edu.wpi.first.apriltag.AprilTagFieldLayout;
+import edu.wpi.first.apriltag.AprilTagFields;
+import edu.wpi.first.math.geometry.Pose2d;
+import edu.wpi.first.math.geometry.Pose3d;
+import edu.wpi.first.networktables.NetworkTableEvent;
+import edu.wpi.first.networktables.NetworkTableInstance;
+import edu.wpi.first.wpilibj.DriverStation;
+import edu.wpi.first.wpilibj.RobotBase;
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+import frc.team2412.robot.Hardware;
+import io.github.oblarg.oblog.annotations.Log;
+import java.io.IOException;
+import java.util.EnumSet;
+import java.util.Optional;
+import java.util.function.BiConsumer;
+import org.photonvision.PhotonCamera;
+import org.photonvision.targeting.PhotonPipelineResult;
+import org.photonvision.targeting.PhotonTrackedTarget;
+
+public class VisionSubsystem extends SubsystemBase {
+	private PhotonCamera photonCamera;
+	private PhotonPipelineResult latestResult;
+	private BiConsumer<Pose2d, Double> poseConsumer;
+	/** Null if no known robot pose, otherwise the last calculated robot pose from vision data. */
+	private Pose3d robotPose = null;
+	/*
+	 * Because we have to handle an IOException, we can't initialize fieldLayout in the variable declaration (private static final AprilTagFieldLayout fieldLayout = ...;). Instead, we have to initialize it in a static initializer (static { ... }).
+	 */
+	private static final AprilTagFieldLayout fieldLayout;
+
+	static {
+		/*
+		 * This code runs when the class is initialized (same time as normal variable initializers).
+		 *
+		 * We have to use temp so that the compiler knows we initialize fieldLayout exactly once. If we directly initialize it, the compiler thinks we might initialize it twice if we get an IOException after initializing it in the try block (since then we'd run the catch block and initialize there too).
+		 */
+		AprilTagFieldLayout temp;
+
+		try {
+			temp = AprilTagFieldLayout.loadFromResource(AprilTagFields.kDefaultField.m_resourceFile);
+		} catch (IOException err) {
+			DriverStation.reportError("Could not load AprilTagFieldLayout! " + err, err.getStackTrace());
+			temp = null;
+		}
+		fieldLayout = temp;
+	}
+
+	public VisionSubsystem(BiConsumer<Pose2d, Double> poseConsumer) {
+		this.poseConsumer = poseConsumer;
+
+		var networkTables = NetworkTableInstance.getDefault();
+
+		// Connect to photonvision server
+		// Only in sim because normally photonvision connects to robot
+		if (RobotBase.isSimulation()) {
+			networkTables.stopServer();
+			networkTables.startClient4("localhost");
+		}
```

Comment:
Does this affect everyone using network tables (`NetworkTableInstance.getDefault()`?). If so, this likely should move to somewhere more common (Robot or Hardware)?

Reply:
It would affect all of NetworkTables, we should test this and the other stuff in https://github.com/orgs/robototes/projects/6?pane=issue&itemId=19585343 tomorrow sometime when drivebase has robot.

Reply:
I tried testing it out today; moving the location had no effect, however, it didnt work properly either way. The sim refuses to connect when without robot, and given we have hummingbonk, I don't think it is worth the effort to implement at this stage